src/libraries/System.Runtime.InteropServices/gen/Microsoft.Interop.SourceGeneration/DiagnosticOr.cs

```diff
@@ -0,0 +1,146 @@
+        {
+            private readonly T _value;
+            internal Val(T value) => _value = value;
+            public override bool IsValue => true;
+            public override T Value => _value;
+            public override Diagnostic Diagnostic => throw new InvalidOperationException();
+        }
+
+        /// <summary>
+        /// Create a Diagnostic variant
+        /// </summary>
+        public static DiagnosticOr<T> From(Diagnostic diagnostic)
+        {
+            Debug.Assert(diagnostic is not null);
+            return new Diag(diagnostic);
+        }
+
+        /// <summary>
+        /// Create a Value variant
+        /// </summary>
+        public static DiagnosticOr<T> From(T value)
+        {
+            Debug.Assert(value is not null);
+            return new Val(value);
+        }
+    }
+
+    public static partial class IncrementalValuesProviderExtensions
+    {
+        /// <summary>
+        /// Splits the elements of <paramref name="provider"/> into a "values" provider and a "diagnositics" provider.
+        /// </summary>
+        public static (IncrementalValuesProvider<T>, IncrementalValuesProvider<Diagnostic>) Split<T>(this IncrementalValuesProvider<DiagnosticOr<T>> provider)
+        {
+            var values = provider.Where(x => x.IsValue).Select(static (x, ct) => x.Value);
+            var diagnostics = provider.Where(x => x.IsDiagnostic).Select(static (x, ct) => x.Diagnostic);
+            return (values, diagnostics);
+        }
+
+        /// <summary>
+        /// Splits the inner arrays of <paramref name="provider"/> into values and diagnostics.
+        /// </summary>
+        public static (IncrementalValuesProvider<SequenceEqualImmutableArray<T>>, IncrementalValuesProvider<Diagnostic>) SplitArrays<T, TEnumerable>(this IncrementalValuesProvider<TEnumerable> provider)
+            where TEnumerable : IEnumerable<DiagnosticOr<T>>
+        {
+            var values = provider.Select((arr, ct) => arr.Where(x => x.IsValue).Select((x, ct) => x.Value).ToSequenceEqualImmutableArray());
+            var diagnostics = provider.SelectMany((arr, ct) => arr.Where(x => x.IsDiagnostic).Select((x, ct) => x.Diagnostic));
+            return (values, diagnostics);
+        }
+
+        /// <summary>
+        /// Splits the elements of <paramref name="provider"/> into groups depending on whether Item1 is a value or a Diagnostic.
+        /// </summary>
+        public static (IncrementalValuesProvider<(T, T2)>, IncrementalValuesProvider<Diagnostic>) Split<T, T2>(this IncrementalValuesProvider<(DiagnosticOr<T>, T2)> provider)
+        {
+            var values = provider.Where(x => x.Item1.IsValue).Select(static (x, ct) => (x.Item1.Value, x.Item2));
+            var diagnostics = provider.Where(x => !x.Item1.IsValue).Select(static (x, ct) => x.Item1.Diagnostic);
+            return (values, diagnostics);
+        }
+    }
+
+    public static partial class IncrementalGeneratorInitializationContextExtensions
```

Comment:
I'd rather have these helper methods in the same file as their corresponding type name instead of making the types partial.

Alternatively, we could make a new class named something like `DiagnosticOrValuesProviderHelpers` where we'd put them.

Reply:
Change the name to DiagnosticOrTHelperExtensions